File: mcp_server_development/node/src/session03/answers/q4-create-server.ts
/**
 * 問題4 の解答：summarize_hours に groupBy（member / project）を追加
 *
 * 問題1 の 3 ツール構成に、集計軸の切り替えを足しています。
 * データ層（data.ts）は変更せず、プロジェクト別の集計はこのファイルに実装します。
 */
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import {
  MAX_RANGE_DAYS,
  daysBetween,
  findMember,
  listMembers,
  projects,
  summarizeHours,
  workLogs,
  type HoursSummary,
  type Member,
  type Project,
} from "../data.js";

const DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;
const MEMBER_ID_PATTERN = /^m-\d{3}$/;

/** 集計軸。スキーマとハンドラでこの 1 か所を共有する */
const GROUP_BY = ["member", "project"] as const;
type GroupBy = (typeof GROUP_BY)[number];

export function createDashboardServerQ4(): McpServer {
  const server = new McpServer({
    name: "team-dashboard",
    version: "0.3.0",
  });

  server.registerTool(
    "list_members",
    {
      title: "メンバー一覧",
      description:
        "チームに所属するメンバーの一覧（ID・氏名・チーム・週の稼働可能時間）を返します。" +
        "稼働時間を集計する前に、有効なメンバー ID を確認する用途で使ってください。",
      inputSchema: {
        team: z
          .enum(["platform", "data"])
          .describe("特定のチームだけに絞る場合に指定します。省略すると全員を返します。")
          .optional(),
      },
    },
    async ({ team }) => ({
      content: [{ type: "text", text: formatMembers(listMembers(team)) }],
    }),
  );

  server.registerTool(
    "list_projects",
    {
      title: "プロジェクト一覧",
      description:
        "プロジェクトの一覧（ID・名称）を返します。" +
        "memberId を指定すると、そのメンバーが稼働記録を持つプロジェクトだけに絞れます。",
      inputSchema: {
        memberId: z
          .string()
          .regex(MEMBER_ID_PATTERN, "m-001 のような形式で指定してください")
          .describe("指定したメンバーが稼働したプロジェクトだけに絞る場合に指定します。")
          .optional(),
      },
    },
    async ({ memberId }) => {
      if (memberId !== undefined && findMember(memberId) === undefined) {
        return toolError(
          `メンバー ID ${memberId} は存在しません。list_members で有効な ID を確認してください。`,
        );
      }
      return { content: [{ type: "text", text: formatProjects(listProjects(memberId)) }] };
    },
  );

  server.registerTool(
    "summarize_hours",
    {
      title: "稼働時間の集計",
      description:
        "指定した期間の稼働時間を集計し、合計と内訳を返します。" +
        "期間は開始日・終了日の両方を含みます。" +
        "groupBy でメンバー別（既定）とプロジェクト別を切り替えられます。" +
        `1 回で集計できるのは最長 ${MAX_RANGE_DAYS} 日です。`,
      inputSchema: {
        from: z
          .string()
          .regex(DATE_PATTERN, "YYYY-MM-DD 形式で指定してください")
          .describe("集計期間の開始日（YYYY-MM-DD、この日を含む）"),
        to: z
          .string()
          .regex(DATE_PATTERN, "YYYY-MM-DD 形式で指定してください")
          .describe("集計期間の終了日（YYYY-MM-DD、この日を含む）"),
        memberId: z
          .string()
          .regex(MEMBER_ID_PATTERN, "m-001 のような形式で指定してください")
          .describe("特定のメンバーだけを集計する場合に指定します。省略すると全員が対象です。")
          .optional(),
        groupBy: z
          .enum(GROUP_BY)
          .describe(
            "集計軸。member はメンバー別、project はプロジェクト別に集計します。省略すると member として扱います。",
          )
          .optional(),
      },
    },
    async ({ from, to, memberId, groupBy }) => {
      const span = daysBetween(from, to);
      if (Number.isNaN(span)) {
        return toolError("from / to には実在する日付を指定してください（例: 2026-08-03）。");
      }
      if (span <= 0) {
        return toolError(`from（${from}）は to（${to}）以前の日付を指定してください。`);
      }
      if (span > MAX_RANGE_DAYS) {
        return toolError(
          `集計できる期間は最長 ${MAX_RANGE_DAYS} 日です（指定された期間は ${span} 日）。` +
            "期間を分けて複数回呼び出してください。",
        );
      }
      if (memberId !== undefined && findMember(memberId) === undefined) {
        return toolError(
          `メンバー ID ${memberId} は存在しません。list_members で有効な ID を確認してください。`,
        );
      }

      // 既定値はスキーマではなくサーバー側で決める
      const mode: GroupBy = groupBy ?? "member";

      switch (mode) {
        case "member": {
          const summary = summarizeHours({ from, to, memberId });
          return { content: [{ type: "text", text: formatSummary(summary) }] };
        }
        case "project": {
          const summary = summarizeByProject({ from, to, memberId });
          return { content: [{ type: "text", text: formatProjectSummary(summary) }] };
        }
        default: {
          // GROUP_BY に値を足してここを書き忘れると、この行がコンパイルエラーになる
          const unreachable: never = mode;
          throw new Error(`未対応の groupBy: ${String(unreachable)}`);
        }
      }
    },
  );

  return server;
}

type ProjectHours = {
  projectId: string;
  name: string;
  totalHours: number;
  memberCount: number;
};

type ProjectSummary = {
  from: string;
  to: string;
  totalHours: number;
  projects: ProjectHours[];
};

/** 期間内の稼働時間をプロジェクト別に集計する */
function summarizeByProject(params: { from: string; to: string; memberId?: string }): ProjectSummary {
  const { from, to, memberId } = params;

  const targets = workLogs.filter(
    (log) =>
      log.date >= from &&
      log.date <= to &&
      (memberId === undefined || log.memberId === memberId),
  );

  const buckets = new Map<string, { hours: number; members: Set<string> }>();
  for (const log of targets) {
    const bucket = buckets.get(log.projectId) ?? { hours: 0, members: new Set<string>() };
    bucket.hours += log.hours;
    bucket.members.add(log.memberId);
    buckets.set(log.projectId, bucket);
  }

  const rows: ProjectHours[] = [];
  for (const [id, bucket] of buckets) {
    rows.push({
      projectId: id,
      name: projects.find((project) => project.id === id)?.name ?? "(不明なプロジェクト)",
      totalHours: roundHours(bucket.hours),
      memberCount: bucket.members.size,
    });
  }
  rows.sort((a, b) => a.projectId.localeCompare(b.projectId));

  return {
    from,
    to,
    totalHours: roundHours(rows.reduce((sum, row) => sum + row.totalHours, 0)),
    projects: rows,
  };
}

function roundHours(hours: number): number {
  return Math.round(hours * 10) / 10;
}

function listProjects(memberId?: string): Project[] {
  const sorted = [...projects].sort((a, b) => a.id.localeCompare(b.id));
  if (memberId === undefined) {
    return sorted;
  }
  const worked = new Set(
    workLogs.filter((log) => log.memberId === memberId).map((log) => log.projectId),
  );
  return sorted.filter((project) => worked.has(project.id));
}

function toolError(message: string) {
  return { content: [{ type: "text" as const, text: message }], isError: true };
}

function formatProjectSummary(summary: ProjectSummary): string {
  const header = `${summary.from} 〜 ${summary.to} の稼働時間: 合計 ${summary.totalHours} 時間 / 対象 ${summary.projects.length} プロジェクト`;
  if (summary.projects.length === 0) {
    return `${header}\n対象期間に稼働記録はありません。`;
  }
  const lines = summary.projects.map(
    (row) => `- ${row.name}（${row.projectId}）: ${row.totalHours} 時間 / ${row.memberCount} 名`,
  );
  return [header, ...lines].join("\n");
}

function formatProjects(found: Project[]): string {
  if (found.length === 0) {
    return "該当するプロジェクトはありません。";
  }
  return [`プロジェクト ${found.length} 件`, ...found.map((p) => `- ${p.id} ${p.name}`)].join("\n");
}

function formatMembers(found: Member[]): string {
  if (found.length === 0) {
    return "該当するメンバーはいません。";
  }
  const lines = found.map(
    (member) =>
      `- ${member.id} ${member.name}（${member.team} / 週 ${member.weeklyCapacityHours} 時間）`,
  );
  return [`メンバー ${found.length} 名`, ...lines].join("\n");
}

function formatSummary(summary: HoursSummary): string {
  const header = `${summary.from} 〜 ${summary.to} の稼働時間: 合計 ${summary.totalHours} 時間 / 対象 ${summary.members.length} 名`;
  if (summary.members.length === 0) {
    return `${header}\n対象期間に稼働記録はありません。`;
  }
  const lines = summary.members.map(
    (row) =>
      `- ${row.name}（${row.memberId} / ${row.team}）: ${row.totalHours} 時間 / ${row.workedDays} 日`,
  );
  return [header, ...lines].join("\n");
}

File: mcp_server_development/python/src/session04/q4_create_server.py
"""問題4（前半）: group_by を追加した独立したサーバー定義

本文のインスタンスに足すのではなく、新しい MCPServer を作っています。
summarize_hours を「差し替える」必要があり、同じ名前のツールを
同じインスタンスに 2 回登録することはできないためです。
"""

from typing import Annotated, Literal

from mcp.server.mcpserver import MCPServer
from pydantic import Field

import data
from create_server import (
    DATE_PATTERN,
    MemberIdStr,
    TeamName,
    ToolFailure,
    format_members,
    format_summary,
)

# 集計の切り口。選択肢を閉じておくと AI が別の値を送ってこない
GroupBy = Literal["member", "project"]

mcp = MCPServer(name="team-dashboard", version="0.1.0")


@mcp.tool(title="メンバー一覧", structured_output=False)
def list_members(
    team: Annotated[
        TeamName | None,
        Field(description="特定のチームだけに絞る場合に指定します。省略すると全員を返します。"),
    ] = None,
) -> str:
    """チームに所属するメンバーの一覧（ID・氏名・チーム・週の稼働可能時間）を返します。稼働時間を集計する前に、有効なメンバー ID を確認する用途で使ってください。"""
    return format_members(data.list_members(team))


@mcp.tool(
    title="稼働時間の集計",
    description=(
        "指定した期間の稼働時間を集計し、合計と内訳を返します。"
        "group_by でメンバー別・プロジェクト別を切り替えられます（既定はメンバー別）。"
        "期間は開始日・終了日の両方を含みます。"
        f"1 回で集計できるのは最長 {data.MAX_RANGE_DAYS} 日です。"
    ),
    structured_output=False,
)
def summarize_hours(
    start_date: Annotated[
        str,
        Field(
            pattern=DATE_PATTERN,
            description="集計期間の開始日（YYYY-MM-DD、この日を含む）",
        ),
    ],
    end_date: Annotated[
        str,
        Field(
            pattern=DATE_PATTERN,
            description="集計期間の終了日（YYYY-MM-DD、この日を含む）",
        ),
    ],
    member_id: Annotated[
        MemberIdStr | None,
        Field(
            description="特定のメンバーだけを集計する場合に指定します。省略すると全員が対象です。",
        ),
    ] = None,
    group_by: Annotated[
        GroupBy | None,
        Field(
            description="集計の切り口。member はメンバー別、project はプロジェクト別。省略するとメンバー別になります。",
        ),
    ] = None,
) -> str:
    """（AI 向けの説明文はデコレータの description= 側に書いています）"""
    span = data.days_between(start_date, end_date)
    if span is None:
        raise ToolFailure(
            "start_date / end_date には実在する日付を指定してください（例: 2026-08-03）。"
        )
    if span <= 0:
        raise ToolFailure(
            f"start_date（{start_date}）は end_date（{end_date}）以前の日付を指定してください。"
        )
    if span > data.MAX_RANGE_DAYS:
        raise ToolFailure(
            f"集計できる期間は最長 {data.MAX_RANGE_DAYS} 日です（指定された期間は {span} 日）。"
            "期間を分けて複数回呼び出してください。"
        )
    if member_id is not None and data.find_member(member_id) is None:
        raise ToolFailure(
            f"メンバー ID {member_id} は存在しません。list_members で有効な ID を確認してください。"
        )

    # 既定値はスキーマに書かず、ここで補う（理由は解説の ②）
    mode: GroupBy = group_by if group_by is not None else "member"

    match mode:
        case "member":
            return format_summary(data.summarize_hours(start_date, end_date, member_id))
        case "project":
            return format_project_summary(start_date, end_date, member_id)
        case _:
            # GroupBy に値を足したのに分岐を書き忘れたときだけここに来る
            raise ToolFailure(f"group_by に未対応の値が指定されました: {mode}")


def summarize_by_project(
    date_from: str, date_to: str, member_id: str | None = None
) -> list[tuple[str, str, float, int]]:
    """プロジェクト別の (ID, 名称, 合計時間, 関わった人数) を ID 昇順で返す"""
    targets = [
        log
        for log in data.WORK_LOGS
        if date_from <= log.date <= date_to
        and (member_id is None or log.member_id == member_id)
    ]

    hours_by_project: dict[str, float] = {}
    people_by_project: dict[str, set[str]] = {}
    for log in targets:
        hours_by_project[log.project_id] = (
            hours_by_project.get(log.project_id, 0) + log.hours
        )
        people_by_project.setdefault(log.project_id, set()).add(log.member_id)

    rows: list[tuple[str, str, float, int]] = []
    for project in sorted(data.PROJECTS, key=lambda project: project.id):
        if project.id not in hours_by_project:
            continue
        rows.append(
            (
                project.id,
                project.name,
                data.round_hours(hours_by_project[project.id]),
                len(people_by_project[project.id]),
            )
        )
    return rows


def format_project_summary(
    date_from: str, date_to: str, member_id: str | None = None
) -> str:
    rows = summarize_by_project(date_from, date_to, member_id)
    total = data.round_hours(sum(row[2] for row in rows))
    header = (
        f"{date_from} 〜 {date_to} の稼働時間: "
        f"合計 {data.format_hours(total)} 時間 / 対象 {len(rows)} プロジェクト"
    )
    if not rows:
        return f"{header}\n対象期間に稼働記録はありません。"
    lines = [
        f"- {name}（{project_id}）: {data.format_hours(hours)} 時間 / {people} 名"
        for project_id, name, hours, people in rows
    ]
    return "\n".join([header, *lines])
